File: frontend/src/lib/domain/summary.ts
/**
 * Day/week aggregation rules — shared by the UI (computing from local
 * entries while offline) and the mock server (computing "server-side"),
 * and mirrored by the Phase-2 backend. One implementation, one answer.
 */

import type { DaySummary, DiaryEntry, Goal, LocalDate, WeeklyReport } from '../api/types';
import { remainingKcal, roundHalfUp, sumNutrients } from './nutrition';
import { addDays, dateRange } from './dates';

export function computeDaySummary(date: LocalDate, entries: DiaryEntry[], goal: Goal): DaySummary {
  const live = entries.filter((e) => e.date === date && !e.deleted);
  const foods = live.filter((e) => e.kind === 'food');
  const consumed = sumNutrients(foods.map((f) => f.nutrients));
  const burnedExercise = live.filter((e) => e.kind === 'exercise').reduce((s, e) => s + e.caloriesBurned, 0);
  const waterMl = live.filter((e) => e.kind === 'water').reduce((s, e) => s + e.amountMl, 0);

  return {
    date,
    consumed,
    burnedExercise,
    waterMl,
    goal: {
      calorieTarget: goal.calorieTarget,
      proteinTargetG: goal.proteinTargetG,
      carbsTargetG: goal.carbsTargetG,
      fatTargetG: goal.fatTargetG,
      waterTargetMl: goal.waterTargetMl,
    },
    remainingKcal: remainingKcal({
      calorieTarget: goal.calorieTarget,
      consumedKcal: consumed.kcal,
      exerciseBurnedKcal: burnedExercise,
      creditExercise: true,
    }),
    entryCount: live.length,
  };
}

export function computeWeeklyReport(weekStart: LocalDate, entries: DiaryEntry[], goal: Goal): WeeklyReport {
  const weekEnd = addDays(weekStart, 6);
  const days = dateRange(weekStart, weekEnd);
  const live = entries.filter((e) => !e.deleted && e.date >= weekStart && e.date <= weekEnd);

  const foodsOn = (date: LocalDate) =>
    live.filter((e): e is Extract<DiaryEntry, { kind: 'food' }> => e.kind === 'food' && e.date === date);

  const perDay = days.map((date) => {
    const kcal = roundHalfUp(foodsOn(date).reduce((s, f) => s + f.nutrients.kcal, 0));
    return { date, kcal, target: goal.calorieTarget };
  });

  const loggedDays = perDay.filter((d) => live.some((e) => e.date === d.date));
  const daysLogged = loggedDays.length;
  const avg = (fn: (d: LocalDate) => number) =>
    daysLogged === 0 ? 0 : roundHalfUp(loggedDays.reduce((s, d) => s + fn(d.date), 0) / daysLogged, 1);

  const weights = live
    .filter((e): e is Extract<DiaryEntry, { kind: 'weight' }> => e.kind === 'weight')
    .sort((a, b) => a.date.localeCompare(b.date));
  const weightChangeKg =
    weights.length >= 2 ? roundHalfUp(weights[weights.length - 1].weightKg - weights[0].weightKg, 1) : null;

  const exercises = live.filter((e) => e.kind === 'exercise');

  const withinTarget = loggedDays.filter((d) => {
    const foodDay = perDay.find((p) => p.date === d.date)!;
    return foodDay.kcal > 0 && Math.abs(foodDay.kcal - goal.calorieTarget) / goal.calorieTarget <= 0.1;
  });

  const bestDay =
    withinTarget.length > 0
      ? withinTarget.reduce((best, d) => {
          const cur = perDay.find((p) => p.date === d.date)!;
          const bst = perDay.find((p) => p.date === best.date)!;
          return Math.abs(cur.kcal - goal.calorieTarget) < Math.abs(bst.kcal - goal.calorieTarget) ? d : best;
        }).date
      : null;

  return {
    weekStart,
    weekEnd,
    avgKcal: avg((d) => foodsOn(d).reduce((s, f) => s + f.nutrients.kcal, 0)),
    avgProteinG: avg((d) => foodsOn(d).reduce((s, f) => s + f.nutrients.proteinG, 0)),
    avgCarbsG: avg((d) => foodsOn(d).reduce((s, f) => s + f.nutrients.carbsG, 0)),
    avgFatG: avg((d) => foodsOn(d).reduce((s, f) => s + f.nutrients.fatG, 0)),
    avgWaterMl: avg((d) =>
      live
        .filter((e): e is Extract<DiaryEntry, { kind: 'water' }> => e.kind === 'water' && e.date === d)
        .reduce((s, e) => s + e.amountMl, 0),
    ),
    daysLogged,
    weightChangeKg,
    exerciseSessions: exercises.length,
    exerciseMinutes: exercises.reduce((s, e) => s + e.durationMin, 0),
    caloriesBurned: roundHalfUp(exercises.reduce((s, e) => s + e.caloriesBurned, 0)),
    adherencePct: daysLogged === 0 ? 0 : roundHalfUp((withinTarget.length / daysLogged) * 100),
    bestDay,
    perDay,
  };
}

File: backend/app/domain/summary.py
"""Day/week aggregation — mirrors frontend/src/lib/domain/summary.ts."""

from __future__ import annotations

from datetime import date as date_type, timedelta
from typing import Any

from .nutrition import remaining_kcal, round_half_up, sum_nutrients


def add_days(day: str, days: int) -> str:
    return (date_type.fromisoformat(day) + timedelta(days=days)).isoformat()


def week_bounds(day: str) -> tuple[str, str]:
    d = date_type.fromisoformat(day)
    start = d - timedelta(days=d.weekday())  # Monday
    return start.isoformat(), (start + timedelta(days=6)).isoformat()


def date_range(start: str, end: str) -> list[str]:
    out, d = [], date_type.fromisoformat(start)
    e = date_type.fromisoformat(end)
    while d <= e:
        out.append(d.isoformat())
        d += timedelta(days=1)
    return out


def compute_day_summary(day: str, entries: list[dict[str, Any]], goal: dict[str, Any]) -> dict[str, Any]:
    """entries: client-shaped dicts (payload merged); goal: camelCase dict."""
    live = [e for e in entries if e["date"] == day and not e.get("deleted")]
    foods = [e for e in live if e["kind"] == "food"]
    consumed = sum_nutrients([f["nutrients"] for f in foods])
    burned = sum(e["caloriesBurned"] for e in live if e["kind"] == "exercise")
    water = sum(e["amountMl"] for e in live if e["kind"] == "water")

    return {
        "date": day,
        "consumed": consumed,
        "burnedExercise": burned,
        "waterMl": water,
        "goal": {
            "calorieTarget": goal["calorieTarget"],
            "proteinTargetG": goal["proteinTargetG"],
            "carbsTargetG": goal["carbsTargetG"],
            "fatTargetG": goal["fatTargetG"],
            "waterTargetMl": goal["waterTargetMl"],
        },
        "remainingKcal": remaining_kcal(goal["calorieTarget"], consumed["kcal"], burned),
        "entryCount": len(live),
    }


def compute_weekly_report(week_start: str, entries: list[dict[str, Any]], goal: dict[str, Any]) -> dict[str, Any]:
    week_end = add_days(week_start, 6)
    days = date_range(week_start, week_end)
    live = [e for e in entries if not e.get("deleted") and week_start <= e["date"] <= week_end]

    def foods_on(d: str) -> list[dict[str, Any]]:
        return [e for e in live if e["kind"] == "food" and e["date"] == d]

    per_day = [
        {"date": d, "kcal": round_half_up(sum(f["nutrients"]["kcal"] for f in foods_on(d))), "target": goal["calorieTarget"]}
        for d in days
    ]

    logged = [p for p in per_day if any(e["date"] == p["date"] for e in live)]
    n = len(logged)

    def avg(fn) -> float:
        return 0 if n == 0 else round_half_up(sum(fn(p["date"]) for p in logged) / n, 1)

    weights = sorted((e for e in live if e["kind"] == "weight"), key=lambda e: e["date"])
    weight_change = round_half_up(weights[-1]["weightKg"] - weights[0]["weightKg"], 1) if len(weights) >= 2 else None

    exercises = [e for e in live if e["kind"] == "exercise"]

    target = goal["calorieTarget"]
    within = [
        p for p in logged
        if p["kcal"] > 0 and abs(p["kcal"] - target) / target <= 0.1
    ]
    best_day = min(within, key=lambda p: abs(p["kcal"] - target))["date"] if within else None

    return {
        "weekStart": week_start,
        "weekEnd": week_end,
        "avgKcal": avg(lambda d: sum(f["nutrients"]["kcal"] for f in foods_on(d))),
        "avgProteinG": avg(lambda d: sum(f["nutrients"].get("proteinG", 0) for f in foods_on(d))),
        "avgCarbsG": avg(lambda d: sum(f["nutrients"].get("carbsG", 0) for f in foods_on(d))),
        "avgFatG": avg(lambda d: sum(f["nutrients"].get("fatG", 0) for f in foods_on(d))),
        "avgWaterMl": avg(lambda d: sum(e["amountMl"] for e in live if e["kind"] == "water" and e["date"] == d)),
        "daysLogged": n,
        "weightChangeKg": weight_change,
        "exerciseSessions": len(exercises),
        "exerciseMinutes": sum(e["durationMin"] for e in exercises),
        "caloriesBurned": round_half_up(sum(e["caloriesBurned"] for e in exercises)),
        "adherencePct": 0 if n == 0 else round_half_up(len(within) / n * 100),
        "bestDay": best_day,
        "perDay": per_day,
    }
